### PR TITLE
Encryption async api

### PR DESCRIPTION
**Pull request by zelig**
This PR modifies the encryption package to prepare it for concurrent segmentwise API

- give encryption key in the constructor so each chunk encryption creates a new encryptor
- abstract out a segmentwise transcrypt method
- it includes padding
- adapt RefEncryption for AC to new API

### CODE REVIEW

File: swarm/storage/encryption/encryption.go [hunk 1]
```diff
@@ -21,91 +21,127 @@ import (
...
+	for i := 0; i < inLength; i += e.keyLen {
+		l := min(e.keyLen, inLength-i)
+		// call transformations per segment (asyncronously)
+		go func(i int, x, y []byte) {
+			defer wg.Done()
+			e.transcrypt(i, x, y)
+		}(i/e.keyLen, in[i:i+l], out[i:i+l])
+	}
+	// pad the rest if out is longer
+	pad(out[inLength:])
```
[the remark this answers is not in the document]
Comment (zelig) [reply swarm/storage/encryption/encryption.go]: well for the async case you can just call `transcrypt` on padded sections with zero length input buffer, but for the syncronous case you need to just pad explicitly

File: swarm/api/encrypt.go [hunk 2]
```diff
@@ -25,27 +25,31 @@ import (
...
+		refSize: refSize,
+		// spanEncryption: encryption.New(0, uint32(refSize/32), sha3.NewKeccak256),
+		// dataEncryption: encryption.New(refSize, 0, sha3.NewKeccak256),
+		span: span,
 	}
 }
 
 func (re *RefEncryption) Encrypt(ref []byte, key []byte) ([]byte, error) {
-	encryptedSpan, err := re.spanEncryption.Encrypt(re.span, key)
+	spanEncryption := encryption.New(key, 0, uint32(re.refSize/32), sha3.NewKeccak256)
```
[the remark this answers is not in the document]
Comment (zelig) [reply swarm/api/encrypt.go]: Yes, we know here the hash is Keccak256

File: swarm/storage/hasherstore.go [hunk 3]
```diff
@@ -201,13 +173,46 @@ func (h *hasherStore) decryptChunkData(chunkData ChunkData, encryptionKey encryp
...
+	}
+	encryptedData, err := h.newDataEncryption(key).Encrypt(chunkData[8:])
+	if err != nil {
+		return nil, nil, err
+	}
+	return encryptedSpan, encryptedData, nil
+}
+
+func (h *hasherStore) newSpanEncryption(key encryption.Key) encryption.Encryption {
+	return encryption.New(key, 0, uint32(chunk.DefaultSize/h.refSize), sha3.NewKeccak256)
```
[the remark this answers is not in the document]
Comment (gbalint) [reply swarm/storage/hasherstore.go]: That is the initial counter value for counter mode encryption, see https://github.com/ethersphere/swarm/wiki/Symmetric-Encryption-for-Swarm-Content and https://en.wikipedia.org/wiki/Block_cipher_mode_of_operation#Counter_(CTR)

[the remark this answers is not in the document]
Comment (nolash) [reply swarm/storage/hasherstore.go]: I am sorry, unclear question. I meant what's behind the particular value.

**Pull request closed without merging**